File: backend/tools/weatherTool.js
/**
 * Weather checking tool
 */
const { getWeatherForDate } = require("../services/weatherService");

/**
 * Check weather for a specific date and optionally time
 * @param {Object} params - Tool parameters
 * @param {string} params.date - Date in YYYY-MM-DD format
 * @param {string} [params.time] - Time in HH:mm format (24-hour, e.g., "19:00" for 7 PM)
 * @param {string} [params.location] - Location (defaults to RESTAURANT_LOCATION or "Mumbai")
 * @returns {Promise<Object>} Weather data with condition, temperature, description
 */
const execute = async (params) => {
  console.log(`🔧 [WEATHER_TOOL] Starting weather check`);
  console.log(`📦 [WEATHER_TOOL] Params:`, JSON.stringify(params, null, 2));

  try {
    const { date, time, location } = params;

    if (!date) {
      const error = "Date parameter is required";
      console.error(`❌ [WEATHER_TOOL] ${error}`);
      throw new Error(error);
    }

    // Validate date format
    if (!/^\d{4}-\d{2}-\d{2}$/.test(date)) {
      const error = "Invalid date format. Use YYYY-MM-DD";
      console.error(`❌ [WEATHER_TOOL] ${error}`);
      throw new Error(error);
    }

    // Validate time format if provided
    if (time && !/^([0-1][0-9]|2[0-3]):[0-5][0-9]$/.test(time)) {
      const error = "Invalid time format. Use 24-hour format (HH:mm)";
      console.error(`❌ [WEATHER_TOOL] ${error}`);
      throw new Error(error);
    }

    const targetLocation =
      location || process.env.RESTAURANT_LOCATION || "Mumbai";
    const timeStr = time ? ` at ${time}` : "";
    console.log(
      `🌤️ [WEATHER_TOOL] Fetching weather for ${date}${timeStr} in ${targetLocation}`
    );

    const weatherData = await getWeatherForDate(date, targetLocation, time);

    console.log(
      `✅ [WEATHER_TOOL] Weather fetched successfully:`,
      JSON.stringify(weatherData, null, 2)
    );

    return {
      success: true,
      data: weatherData,
    };
  } catch (error) {
    console.error(`❌ [WEATHER_TOOL] Error:`, error.message);
    console.error("Stack trace:", error.stack);
    return {
      success: false,
      error: error.message || "Failed to fetch weather data",
    };
  }
};

module.exports = {
  name: "weather",
  description: "Check weather forecast for a specific date",
  execute,
};

File: agent/tools/weather_tool.py
"""
Weather checking tool wrapper
"""
import re
from .base_tool import BaseTool
from utils.api_client import BackendAPIClient
from typing import Dict, Any


class WeatherTool(BaseTool):
    """Tool for checking weather forecast"""
    
    def __init__(self):
        super().__init__(
            name="weather",
            description="Check weather forecast for a specific date"
        )
        self.api_client = BackendAPIClient()
    
    async def execute(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """
        Check weather for a date and optionally time
        Args:
            params: {'date': 'YYYY-MM-DD', 'time': 'HH:mm' (optional)}
        Returns:
            Weather data dict
        """
        print(f"🔧 [WEATHER_TOOL] Executing with params: {params}")
        error = self.validate_params(params, ['date'])
        if error:
            print(f"❌ [WEATHER_TOOL] Validation failed: {error}")
            return {'success': False, 'error': error}
        
        # Validate time format if provided
        if 'time' in params and params['time']:
            if not re.match(r'^([0-1][0-9]|2[0-3]):[0-5][0-9]$', params['time']):
                error = "Invalid time format. Use 24-hour format (HH:mm)"
                print(f"❌ [WEATHER_TOOL] {error}")
                return {'success': False, 'error': error}
        
        try:
            result = await self.api_client.call_tool('weather', params)
            print(f"✅ [WEATHER_TOOL] Weather data retrieved")
            return result
        except Exception as e:
            print(f"❌ [WEATHER_TOOL] Error: {str(e)}")
            import traceback
            traceback.print_exc()
            return {
                'success': False,
                'error': str(e)
            }
